### PR TITLE
X11 and OpenGL(ES) renderer cleanup

### PR DESCRIPTION
**Pull request by MaxKellermann**

### CODE REVIEW

**Comment by MaxKellermann** [comment]
Fixed the cmake build failure.

**Comment by FernetMenta** [comment]
merging RenderSystemGL and RenderSystemGLES and introducing ifdeffery is not a good idea.

**Comment by MaxKellermann** [comment]
There's a lot of redundant code, and redundant code leads to bugs being fixed only in one copy, and copies eventually running out of sync. I agree that ifdefs are ugly, but duplicate code is even more ugly.

**Comment by FernetMenta** [comment]
I don't agree. Merging those just to reduce duplicate code won't happen

**Comment by MaxKellermann** [comment]
Ok, last commit removed. I'll keep it in my fork, though.

Do you have a better idea how to make windowing implementations less aware of the GL variant in use? For example, X11 supports both, but requires some ifdeffery for that - my idea is to move the ifdefs to one central locations, and make windowing implementations not care at all. So we're just talking about **where** the ifdefs live, not **whether** they exist.

**Comment by FernetMenta** [comment]
> Do you have a better idea how to make windowing implementations less aware of the GL variant in use? 

similar to this:
https://github.com/xbmc/xbmc/tree/master/xbmc/cores/VideoPlayer/Process

Instead of cluttering the code with ifdefs, overrides can be processed at build time.

**Comment by MaxKellermann** [comment]
Does this code even work? There is exactly one "override", but CProcessInfoPi::GetFallbackDeintMethod() doesn't actually override anything - because CProcessInfo::GetFallbackDeintMethod() is not virtual.

So your way of doing this **still requires ifdef** and adds runtime overhead.

**Comment by FernetMenta** [comment]
As mentioned, the compiler processes the override. Hence no need to make anything virtual. With this technique a port can be created without touching mainstream code.

**Comment by MaxKellermann** [comment]
No, I don't think that's true. Maybe I misunderstand completely how the C++ compiler works, but in my opinion, this technique is flawed and doesn't work at all. Why would the compiler process this override?

For example, there's CDVDVideoCodecFFmpeg::SetFilters() which calls m_processInfo.GetFallbackDeintMethod(). m_processInfo is a "CProcessInfo&". So "m_processInfo.GetFallbackDeintMethod()" will invoke CProcessInfo::GetFallbackDeintMethod(), without a vtable lookup - because it's not virtual.

Why on earth would the compiler replace CProcessInfo::GetFallbackDeintMethod() with CProcessInfoPi::GetFallbackDeintMethod()? How does the compiler at this point even know that class CProcessInfoPi exists? That doesn't make any sense for me. What if there are other overrides? Which one wins?

Can you please point me to some documentation or description of this technique? If I'm wrong, I'd like to learn about this.

**Comment by MaxKellermann** [comment]
I was really curious about this misunderstanding, so I compiled a `kodi.bin` with the ProcessInfoPi override. This is from the disassembly of CDVDVideoCodecFFmpeg::SetFilters():

    0x0000000000a7fb31 <+165>:   callq  0xb46eba <CProcessInfo::GetFallbackDeintMethod()>

Doesn't look like the override worked, but let's verify if there were linker tricks which replaced CProcessInfo::GetFallbackDeintMethod() with CProcessInfoPi::GetFallbackDeintMethod():

    Dump of assembler code for function CProcessInfo::GetFallbackDeintMethod():
       0x0000000000b46eba <+0>:     mov    $0x7,%eax
       0x0000000000b46ebf <+5>:     retq   
    End of assembler dump.

This is VS_INTERLACEMETHOD_DEINTERLACE=7, source code:

    EINTERLACEMETHOD CProcessInfo::GetFallbackDeintMethod()
    {
      return VS_INTERLACEMETHOD_DEINTERLACE;
    }

Let's look at CProcessInfoPi::GetFallbackDeintMethod() (where `VS_INTERLACEMETHOD_DEINTERLACE_HALF=16`):

    EINTERLACEMETHOD CProcessInfoPi::GetFallbackDeintMethod()
    {
      return VS_INTERLACEMETHOD_DEINTERLACE_HALF;
    }

Disassembly:

    Dump of assembler code for function CProcessInfoPi::GetFallbackDeintMethod():
       0x0000000000b46f80 <+0>:     mov    $0x10,%eax
       0x0000000000b46f85 <+5>:     retq   
    End of assembler dump.

Bottom line: CProcessInfoPi::GetFallbackDeintMethod() is not used, and this override doesn't work, just as I assumed.

Is this proof enough that you're wrong? Or did I do something wrong? Does your disassembly look different?

**Comment by MaxKellermann** [comment]
I forgot to paste the proof that the ProcessInfoPi "override" is really enabled:

    Dump of assembler code for function CProcessInfo::CreateInstance():
       0x0000000000b46f46 <+0>:     push   %rbp
       0x0000000000b46f47 <+1>:     push   %rbx
       0x0000000000b46f48 <+2>:     sub    $0x8,%rsp
       0x0000000000b46f4c <+6>:     mov    $0x8,%edi
       0x0000000000b46f51 <+11>:    callq  0x71a980 <_Znwm@plt>
       0x0000000000b46f56 <+16>:    mov    %rax,%rbx
       0x0000000000b46f59 <+19>:    mov    %rax,%rdi
       0x0000000000b46f5c <+22>:    callq  0xb46f14 <CProcessInfoPi::CProcessInfoPi()>
       0x0000000000b46f61 <+27>:    jmp    0xb46f76 <CProcessInfo::CreateInstance()+48>
       0x0000000000b46f63 <+29>:    mov    %rax,%rbp
       0x0000000000b46f66 <+32>:    mov    %rbx,%rdi
       0x0000000000b46f69 <+35>:    callq  0x71a970 <_ZdlPv@plt>
       0x0000000000b46f6e <+40>:    mov    %rbp,%rdi
       0x0000000000b46f71 <+43>:    callq  0x71a910 <_Unwind_Resume@plt>
       0x0000000000b46f76 <+48>:    mov    %rbx,%rax
       0x0000000000b46f79 <+51>:    add    $0x8,%rsp
       0x0000000000b46f7d <+55>:    pop    %rbx
       0x0000000000b46f7e <+56>:    pop    %rbp
       0x0000000000b46f7f <+57>:    retq   
    End of assembler dump.

As you see, this instantiates CProcessInfoPi, not CProcessInfo. Which doesn't matter, because all callers access this instance through the CProcessInfo interface, and don't know about CProcessInfoPi.

**Comment by MaxKellermann** [comment]
Yet more proof: I downloaded http://milhouse.openelec.tv/builds/master/RPi/OpenELEC-RPi.arm-8.0-Milhouse-20160319000101-%230318-g0026734.tar and disassembled `kodi.bin`. I assume that OpenELEC's Raspberry Pi binaries are sufficiently and properly configured to use Raspberry Pi features, aren't they?

    Dump of assembler code for function _ZN20CDVDVideoCodecFFmpeg10SetFiltersEv:
    [...]
   0x004a1b2c <+40>:    bl      0x4ca224 <_ZN12CProcessInfo22GetFallbackDeintMethodEv>

As you see, CProcessInfo::GetFallbackDeintMethod() is used:

    Dump of assembler code for function _ZN12CProcessInfo22GetFallbackDeintMethodEv:
       0x004ca224 <+0>:     mov     r0, #7
       0x004ca228 <+4>:     bx      lr
    End of assembler dump.

.. and not CProcessInfoPi::GetFallbackDeintMethod():

    Dump of assembler code for function _ZN14CProcessInfoPi22GetFallbackDeintMethodEv:
       0x004ca2d8 <+0>:     mov     r0, #16
       0x004ca2dc <+4>:     bx      lr
    End of assembler dump.

Do you believe me now?

**Comment by fritsch** [comment]
@MaxKellermann give it a bit time - you are right with what you say. Thanks for the detailed explanation and I learned something :-). 

For how to organize the code (besides the bugs you found obviously) it is still fernet's decision, so let's wait a bit for his answer.

**Comment by FernetMenta** [comment]
Thanks for the other PR. virtual indeed was missing. I thought this was using complete substitution.

Nevertheless the concept stays the same: instead of having lots of ifdefs in mainstream, a platform port should override or substitute defined parts of mainline.
https://github.com/xbmc/xbmc/commit/d0364aaa64ee1f97f1424e13330b6c249c6eb487 shows an issue related to platform ports. X11/GLES is no reference platform and we don't even do CI testing for it. It stayed broken for a long time. IMO much better is to have port on their own branch and sync with mainstream when desired.

**Comment by MaxKellermann** [comment]
Commit d0364aaa64ee1f97f1424e13330b6c249c6eb487 is actually a good example for why we should merge those different GL/GLES libraries - this breakage would not have happened, it would have been noticed earlier. Or better: the GLES library would not have required extra fixing, because there would be no GLES library.

The code I have touched here does not actually follow the design you just described: according to your description, there would be no duplicate code - only arch-specific libraries containing the deltas. What we currently have is large chunks of duplicated code with minor differences, with no deltas.

My intention was to merge duplicate code, and merge with other duplicate code of other platforms - for example Android uses EGL, but shares no code with X11-EGL. Both are "reference platforms".

**Comment by MaxKellermann** [comment]
I've rewritten of the GL/GLES merge commit. @FernetMenta please check if this meets your desired design. There's now a common base class, and the GL/GLES derivations only contain the differences.

**Comment by FernetMenta** [comment]
Even if there is similar or duplicate code in GL/GLES, those are different platforms. I don't consider duplicate code an issue. I don't want to care about GLES when I do changes to GL and vice versa. Take this PR as an example: It would require testing on all platforms but Windows despite it is platform code.
This is the opposite direction I want to go and we already discussed this. I don't want to create the same mess as in windowing/EGL.

**Comment by MaxKellermann** [comment]
OK, I have removed both commits which deal with duplicate code, even though I believe your decision leads to inferior code quality. I can't comprehend why you believe that OpenGL and GLES are different platforms. GLES was designed with compatibility in mind; it is easy to write code which is compatible with both.

Currently, there are 30 code sections in Kodi surrounded by "#if defined(HAS_GL) || defined(HAS_GLES)", i.e. code is shared between OpenGL and GLES (which is a good thing according to my philosophy, and bad according to yours). Quite a few of those have additional #ifdefs to implement the minor differences between the two APIs. These 30 code locations are just as wrong as the code I submitted, aren't they? Will you accept PRs which duplicates the code, to eliminate those #ifdefs? Or is there something I misunderstood in your design?

**Comment by FernetMenta** [comment]
> Or is there something I misunderstood in your design?

The idea of the ProcessInfo (this class is at its very beginning) design is to allow a platform or even a variant of an existing platform to change this without touching mainline code. A platform port can be done by placing new files in the overrides folder. I don't claim that it is perfect at this stage but this is a start. This is a proven concept that is not new and not invented by myself.

A hierarchy of platform ports somehow conflicts with this. This would be like an addon to an addon.

jenkins build this please
